gnocchi/rest/aggregates/api.py

```diff
@@ -0,0 +1,209 @@
+import pyparsing
+import six
+import voluptuous
+
+from gnocchi.rest.aggregates import operations as agg_operations
+from gnocchi.rest.aggregates import processor
+from gnocchi.rest import api
+from gnocchi import storage
+from gnocchi import utils
+
+
+def validate_qs(start, stop, granularity, needed_overlap, fill):
+    try:
+        needed_overlap = float(needed_overlap)
+    except ValueError:
+        api.abort(400, 'needed_overlap must be a number')
+    if needed_overlap != 100.0 and start is None and stop is None:
+        api.abort(400, 'start and/or stop must be provided if specifying '
+                  'needed_overlap')
+
+    if start is not None:
+        try:
+            start = utils.to_timestamp(start)
+        except Exception:
+            api.abort(400, "Invalid value for start")
+
+    if stop is not None:
+        try:
+            stop = utils.to_timestamp(stop)
+        except Exception:
+            api.abort(400, "Invalid value for stop")
+
+    if granularity is not None:
+        try:
+            granularity = utils.to_timespan(granularity)
+        except ValueError as e:
+            api.abort(400, e)
+
+    if fill is not None:
+        if granularity is None:
+            api.abort(400, "Unable to fill without a granularity")
+        if fill != "null":
+            try:
+                fill = float(fill)
+            except ValueError as e:
+                api.abort(400, "fill must be a float or \'null\': %s" % e)
+    return start, stop, granularity, needed_overlap, fill
+
+
+def _OperationsSubNodeSchema(v):
+    return OperationsSubNodeSchema(v)
+
+
+def MetricSchema(v):
+    """metric keyword schema
+
+    It could be:
+
+    ["metric", "metric-ref", "aggregation"]
+
+    or
+
+    ["metric, ["metric-ref", "aggregation"], ["metric-ref", "aggregation"]]
+    """
+    if not isinstance(v, (list, tuple)) or len(v) <= 2 or v[0] != u"metric":
+        raise voluptuous.Invalid("'metric' is invalid")
+    return [u"metric"] + voluptuous.Schema(voluptuous.Any(
+        voluptuous.ExactSequence([six.text_type, six.text_type]),
+        voluptuous.All(
+            voluptuous.Length(min=1),
+            [voluptuous.ExactSequence([six.text_type, six.text_type])],
+        )), required=True)(v[1:])
+
+
+OperationsSchemaBase = [
+    MetricSchema,
+    voluptuous.ExactSequence(
+        [u"aggregate",
+         voluptuous.Any(*list(agg_operations.AGG_MAP.keys())),
+         _OperationsSubNodeSchema]
+    ),
+]
+
+
+OperationsSubNodeSchema = voluptuous.Schema(voluptuous.Any(*tuple(
+    OperationsSchemaBase + [voluptuous.Coerce(float)]
+)), required=True)
+
+
+def OperationsSchema(v):
+    if isinstance(v, six.text_type):
+        try:
+            v = pyparsing.OneOrMore(
+                pyparsing.nestedExpr()).parseString(v).asList()[0]
+        except pyparsing.ParseException as e:
+            raise ValueError("Invalid value for operations: %s" %
+                             six.text_type(e))
```

Comment:
This does not look super user friendly. It's likely this is what is gonna be returned to the user using the CLI. Can we have a better error message with pyparsing? :(

Reply:
I have abort() directly here to show a detailed 400, six.text_type(e) produce a understandable error.